coprocessor/fhevm-engine/host-listener/tests/host_listener_integration_tests.rs

```diff
@@ -522,6 +561,139 @@ async fn test_catchup_only() -> Result<(), anyhow::Error> {
+    _setup: Setup,
+    listener_handle: tokio::task::JoinHandle<anyhow::Result<()>>,
+    tfhe_events_count: i64,
+    acl_events_count: i64,
+    nb_wallets: i64,
+}
+
+async fn run_catchup_only_scenario<F>(
+    nb_event_per_wallet: i64,
+    sleep_secs: u64,
+    configure_args: F,
+) -> Result<CatchupOutcome, anyhow::Error>
+where
+    F: FnOnce(&mut Args),
+{
+    let setup = setup(None).await?;
+    let mut args = setup.args.clone();
+
+    let wallets_clone = setup.wallets.clone();
+    let url_clone = setup.args.url.clone();
+    let tfhe_contract_clone = setup.tfhe_contract.clone();
+    let acl_contract_clone = setup.acl_contract.clone();
+    emit_events(
+        &wallets_clone,
+        &url_clone,
+        tfhe_contract_clone,
+        acl_contract_clone,
+        false,
+        nb_event_per_wallet,
+    )
+    .await;
+
+    configure_args(&mut args);
+    args.only_catchup_loop = true;
+
+    let listener_handle = tokio::spawn(main(args.clone()));
+    assert!(health_check::wait_healthy(&setup.health_check_url, 60, 1).await);
+    tokio::time::sleep(tokio::time::Duration::from_secs(sleep_secs)).await;
+
+    let tfhe_events_count = sqlx::query!("SELECT COUNT(*) FROM computations")
+        .fetch_one(&setup.db_pool)
+        .await?
+        .count
+        .unwrap_or(0);
+    let acl_events_count = sqlx::query!("SELECT COUNT(*) FROM allowed_handles")
+        .fetch_one(&setup.db_pool)
+        .await?
+        .count
+        .unwrap_or(0);
+    let nb_wallets = setup.wallets.len() as i64;
+
+    Ok(CatchupOutcome {
+        _setup: setup,
+        listener_handle,
+        tfhe_events_count,
+        acl_events_count,
+        nb_wallets,
+    })
+}
+
+#[tokio::test]
+#[serial(db)]
+async fn test_catchup_only_absolute_end() -> Result<(), anyhow::Error> {
+    let nb_event_per_wallet = 5;
+    let outcome = run_catchup_only_scenario(nb_event_per_wallet, 15, |args| {
+        args.start_at_block = Some(-50);
+        args.end_at_block = Some(50);
+        args.catchup_sleep_secs = 5;
+        args.catchup_paging = 10;
+    })
+    .await?;
+
+    assert_eq!(
+        outcome.tfhe_events_count,
+        outcome.nb_wallets * nb_event_per_wallet
+    );
+    assert_eq!(
+        outcome.acl_events_count,
+        outcome.nb_wallets * nb_event_per_wallet
+    );
+
+    // Listener should still be running (it's in a loop, sleeping between iterations)
+    assert!(
+        !outcome.listener_handle.is_finished(),
+        "Listener should continue running in loop mode"
+    );
+
+    outcome.listener_handle.abort();
+    Ok(())
+}
+
+#[tokio::test]
+#[serial(db)]
+async fn test_catchup_only_relative_end() -> Result<(), anyhow::Error> {
+    let nb_event_per_wallet = 5;
+    let outcome = run_catchup_only_scenario(nb_event_per_wallet, 15, |args| {
+        args.start_at_block = Some(-50); // 50 blocks from current
+        args.end_at_block = Some(-5); // 5 blocks from current (more recent)
```

Comment:
how does it succeed with negative end_at_block ? is it looping or end was not detected ?

Reply:
maybe it should check that it's doing a new catchup aftert the first one